File: frontend/src/features/bridge/retry.ts
import { isRetryable } from './errors';

export interface RetryAdvice {
  retry: boolean;
  afterSeconds: number;
  reason: string;
}

export function adviceFor(code: string, attempt = 0, retryAfter?: number): RetryAdvice {
  if (!isRetryable(code)) {
    return { retry: false, afterSeconds: 0, reason: code };
  }
  if (attempt >= 4) {
    return { retry: false, afterSeconds: 0, reason: 'exhausted' };
  }
  const base = 0.4 * 2 ** attempt;
  const wait = retryAfter !== undefined ? retryAfter : Math.min(8, base);
  return { retry: true, afterSeconds: wait, reason: code };
}

export function parseRetryAfter(header: string | null | undefined): number | undefined {
  if (!header) return undefined;
  const value = Number(header);
  if (!Number.isFinite(value) || value < 0) return undefined;
  return Math.min(value, 60);
}

export async function sleep(seconds: number): Promise<void> {
  if (seconds <= 0) return;
  await new Promise((resolve) => {
    setTimeout(resolve, seconds * 1000);
  });
}

File: src/integration/retry.py
"""Retry / backoff policy for the UI client and gateway rate limiter."""

from __future__ import annotations

from dataclasses import dataclass

from src.integration.errors import ErrorCode


@dataclass(frozen=True, slots=True)
class RetryAdvice:
    retry: bool
    after_seconds: float
    reason: str


_RETRYABLE = frozenset({ErrorCode.RATE_LIMITED, ErrorCode.UNAVAILABLE, ErrorCode.INTERNAL})


def advice_for(
    code: ErrorCode, *, attempt: int = 0, retry_after: float | None = None
) -> RetryAdvice:
    if code not in _RETRYABLE:
        return RetryAdvice(retry=False, after_seconds=0.0, reason=code.value)
    attempt_n = max(0, int(attempt))
    if attempt_n >= 4:
        return RetryAdvice(retry=False, after_seconds=0.0, reason="exhausted")
    base = 0.4 * (2**attempt_n)
    wait = float(retry_after) if retry_after is not None else min(8.0, base)
    return RetryAdvice(retry=True, after_seconds=wait, reason=code.value)


def parse_retry_after(header: object) -> float | None:
    raw = str(header or "").strip()
    if not raw:
        return None
    try:
        value = float(raw)
    except ValueError:
        return None
    if value < 0:
        return None
    return min(value, 60.0)
